### PR TITLE
feat(monorepo): `yarn` -> `pnpm` migration

### PR DESCRIPTION
**Pull request by clabby**
# Overview

> **Note**
> Very broken, WIP.

Migrates the monorepo from `yarn` -> `pnpm`.

### CODE REVIEW

**Comment by clabby** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5260** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/5260" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/5260?utm_source=stack-comment).

File: packages/hardhat-deploy-config/README.md [hunk 1]
```diff
@@ -5,5 +5,5 @@
 ## Installation
 
 ```
-yarn add @eth-optimism/hardhat-deploy-config
+pnpm add @eth-optimism/hardhat-deploy-config
```
Comment (roninjin10) [inline packages/hardhat-deploy-config/README.md:8]:
```suggestion
pnpm i @eth-optimism/hardhat-deploy-config
```

**Comment by roninjin10** [comment]
If we are using resolutions or hoist in our package.json those no longer are doing anything and you can delete.

You can delete the workspaces key from package.json

Here is dangerously hoist docs if you need it https://pnpm.io/npmrc#shamefully-hoist (though hoisting stuff that we previously said no hoist will break things)

File: .circleci/config.yml [hunk 2]
```diff
@@ -67,15 +67,15 @@ jobs:
       - check-changed:
           patterns: op-bindings,op-chain-ops,packages/
       - restore_cache:
-          name: Restore Yarn Package Cache
+          name: Restore PNPM Package Cache
           keys:
-            - yarn-packages-v2-{{ checksum "yarn.lock" }}
+            - pnpm-packages-v2-{{ checksum "pnpm-lock.yaml" }}
```
[the remark this answers is not in the document]
Comment (clabby) [reply .circleci/config.yml:72]: Should be installed through the ci-builder though, no?

File: .npmrc [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+hoist-pattern[]=!@eth-optimism/contracts-bedrock/@rari-capital/solmate
+hoist-pattern[]=!@openzeppelin
+hoist-pattern[]=!@eth-optimism/contracts-periphery/ds-test
+hoist-pattern[]=!@eth-optimism/contracts-periphery/forge-std
+hoist-pattern[]=!@eth-optimism/contracts-periphery/@rari-capital/solmate
+hoist-pattern[]=!forta-agent
+
+# run prefoo scripts
+# npm and yarn support this as a default pnpm defaults to false
+enable-pre-post-scripts=true
```
Comment (roninjin10) [inline .npmrc:14]: We should remove this behavior [in favor of nx](https://github.com/ethereum-optimism/optimism/pull/4778) controlling the order that things need to be built

File: .npmrc [hunk 4]
```diff
@@ -0,0 +1,14 @@
+hoist-pattern[]=!**/typechain/*
+hoist-pattern[]=!**/@typechain/*
+hoist-pattern[]=!@eth-optimism/contracts-bedrock/ds-test
+hoist-pattern[]=!@eth-optimism/contracts-bedrock/forge-std
+hoist-pattern[]=!@eth-optimism/contracts-bedrock/@rari-capital/solmate
+hoist-pattern[]=!@openzeppelin
+hoist-pattern[]=!@eth-optimism/contracts-periphery/ds-test
+hoist-pattern[]=!@eth-optimism/contracts-periphery/forge-std
+hoist-pattern[]=!@eth-optimism/contracts-periphery/@rari-capital/solmate
+hoist-pattern[]=!forta-agent
```
Comment (roninjin10) [inline .npmrc:10]:
```suggestion
``` 
Haven't tested yet but these are all unnecessary.  By default, pnpm never hoists.   This makes a lot of issues I've seen @maurelian and @smartcontracts battle with from time to time go away.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -43,7 +43,7 @@ You *must* include a `changeset` file in your PR when making a change that would
 Adding a `changeset` file is easy:
 
 1. Navigate to the root of the monorepo.
-2. Run `yarn changeset`. You'll be prompted to select packages to include in the changeset. Use the arrow keys to move the cursor up and down, hit the `spacebar` to select a package, and hit `enter` to confirm your selection. Select *all* packages that require a new release as a result of your PR.
+2. Run `pnpm changeset`. You'll be prompted to select packages to include in the changeset. Use the arrow keys to move the cursor up and down, hit the `spacebar` to select a package, and hit `enter` to confirm your selection. Select *all* packages that require a new release as a result of your PR.
```
Comment (roninjin10) [inline CONTRIBUTING.md:46]: Unrelated to this pr but this is too much detail.  Once you run pnpm changeset you can figure out the rest because the prompt guides you along

File: integration-tests/package.json [hunk 6]
```diff
@@ -4,8 +4,8 @@
   "version": "0.5.22",
   "description": "[Optimism] Integration tests",
   "scripts": {
-    "lint": "yarn lint:fix && yarn lint:check",
-    "lint:fix": "yarn lint:check --fix",
+    "lint": "pnpm lint:fix && pnpm lint:check",
```
Comment (roninjin10) [inline integration-tests/package.json:7]: Unrelated to this pr but this is not a normal way of doing things.   CI should be using lint:check and this has no reason to be running the linter twice

```suggestion
    "lint": "pnpm lint:fix",
```

Don't commit this as this is out of scope for pr

File: ops/docker/Dockerfile.packages [hunk 7]
```diff
@@ -17,68 +17,75 @@ FROM alpine:3.16 as manifests
 RUN apk add coreutils
 
 WORKDIR /tmp
-COPY yarn.lock .nvmrc package.json ./src/
+COPY pnpm-lock.yaml pnpm-workspace.yaml .nvmrc package.json ./src/
```
Comment (roninjin10) [inline ops/docker/Dockerfile.packages:20]: @clabby I added pnpm-workspace.yaml here

File: ops/docker/Dockerfile.packages [hunk 8]
```diff
@@ -17,68 +17,75 @@ FROM alpine:3.16 as manifests
...
   # .nvmrc
   cp .nvmrc ../manifests/
 
 FROM ethereumoptimism/foundry:latest as foundry
-FROM node:16-alpine3.14 as base
+# bullseye-slim is debian based
+# we use it rather than alpien because it's not much
+# bigger and alpine is often missing packages for node applications
+# alpine is not officially supported by node.js
+FROM node:16.16.0-bullseye-slim as base
```
Comment (roninjin10) [inline ops/docker/Dockerfile.packages:38]: This should be in it's own PR and reviewed by zach and any security folks.   This is a pretty big change to use debian instead of alpine.

From my experience and also gathering the wisdom of the crowd building js docker images, because node.js doesn't support alpine officially and doesn't recomend using it, and because I and many others have struggled in past mantaining alpine builds, it's not worth using alpine. 

Negative tradeoff

- bullseye-slim is less minimal than alpine so slightly bigger surface area for security vulnerabilities.   It's an extremely common and battle tested image though
- bullseye-slim is bigger.  It's not significantly bigger though and that is offset a bit by doing a multistage build

File: ops/docker/Dockerfile.packages [hunk 9]
```diff
@@ -17,68 +17,75 @@ FROM alpine:3.16 as manifests
...
 
 # Base: install deps
-RUN apk --no-cache add curl \
+RUN apt-get update && apt-get install -y \
+  curl \
   jq \
   python3 \
   ca-certificates \
   git \
+  g++ \
```
Comment (roninjin10) [inline ops/docker/Dockerfile.packages:47]: This is needed to compile c++ when running pnpm install

File: ops/docker/Dockerfile.packages [hunk 10]
```diff
@@ -17,68 +17,75 @@ FROM alpine:3.16 as manifests
...
   make \
   gcc \
+  pkg-config \
```
Comment (roninjin10) [inline ops/docker/Dockerfile.packages:50]:
```suggestion
```
TODO I accidentally added this twice

File: ops/docker/Dockerfile.packages [hunk 11]
```diff
@@ -17,68 +17,75 @@ FROM alpine:3.16 as manifests
...
   musl-dev \
-  linux-headers \
   bash \
-  build-base \
-  gcompat
```
Comment (roninjin10) [inline ops/docker/Dockerfile.packages:46]: These 3 deps I deleted are needed on the ultra minimal alpine image but bullseye-slim has support for this functionality already.    These are all needed to create compatibility between alpine and c++ compiler on alpine but not debian

File: ops/docker/Dockerfile.packages [hunk 12]
```diff
@@ -17,68 +17,75 @@ FROM alpine:3.16 as manifests
...
+  # the following 4 deps are needed for node-hid
+  # which is a deep sub dependency of ethers to install
+  # correctly
+  pkg-config \
+  libusb-1.0-0-dev \
+  libudev-dev \
+  --no-install-recommends
```
Comment (roninjin10) [inline ops/docker/Dockerfile.packages:59]: a very problematic dependency in my experience is node-hid which is a deep sub dependency of ethers.    Here these installations simply allow some postinstall scripts to run correctly.  yarn would allow these to fail silently

File: ops/docker/Dockerfile.packages [hunk 13]
```diff
@@ -17,68 +17,75 @@ FROM alpine:3.16 as manifests
...
 
-ENV GLIBC_KEY=https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub
-ENV GLIBC_KEY_FILE=/etc/apk/keys/sgerrand.rsa.pub
-ENV GLIBC_RELEASE=https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.35-r0/glibc-2.35-r0.apk
-
-RUN wget -q -O ${GLIBC_KEY_FILE} ${GLIBC_KEY} \
-  && wget -O glibc.apk ${GLIBC_RELEASE} \
-  && apk add glibc.apk --force
```
Comment (roninjin10) [inline ops/docker/Dockerfile.packages:54]: We don't need to do this anymore because it's debian and not alpine and this works by default

File: ops/docker/Dockerfile.packages [hunk 14]
```diff
@@ -17,68 +17,75 @@ FROM alpine:3.16 as manifests
...
+RUN npm install pnpm --global
```
Comment (roninjin10) [inline ops/docker/Dockerfile.packages:61]: @clabby this is why the docker build wasn't working

File: ops/docker/Dockerfile.packages [hunk 15]
```diff
@@ -17,68 +17,75 @@ FROM alpine:3.16 as manifests
...
 
 # Copy manifest files into the image in
-# preparation for `yarn install`.
+# preparation for `pnpm install`.
 COPY integration-tests/package.json ./integration-tests/package.json
 COPY --from=manifests /tmp/manifests  ./
 COPY *.json ./
 
-RUN yarn install --frozen-lockfile && yarn cache clean
+RUN pnpm install
```
Comment (roninjin10) [inline ops/docker/Dockerfile.packages:74]:
```suggestion
RUN pnpm install --frozen-lockfile
```
TODO I messed this up this should still frozen lockfile

File: ops/docker/ci-builder/Dockerfile [hunk 16]
```diff
@@ -49,14 +49,19 @@ RUN apt-get update && \
...
   go install gotest.tools/gotestsum@latest && \
   curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(go env GOPATH)/bin v1.48.0 && \
   curl -fLSs https://raw.githubusercontent.com/CircleCI-Public/circleci-cli/master/install.sh | bash && \
   chmod +x /usr/local/bin/check-changed
 
+RUN echo "downloading pnpm" && \
+  curl -o /pnpm-install.sh -L https://get.pnpm.io/install.sh && \
+  chmod +x /pnpm-install.sh && \
+  /pnpm-install.sh && \
+  rm /pnpm-install.sh
```
Comment (roninjin10) [inline ops/docker/ci-builder/Dockerfile:63]: Could just `npm i -g pnpm` instead

File: ops/docker/hardhat/Dockerfile [hunk 17]
```diff
@@ -4,11 +4,17 @@ FROM node:16-alpine
...
 # install deps
 WORKDIR /hardhat
-RUN yarn install && yarn cache clean
+
+RUN npm install pnpm --global
+# Note this is not using the pnpm lockfile but rather installing from scratch
+# this is not deterministic and the versions being installed here are not guaranteed to match
+# the versions in the lockfile.
+# be aware of this when debugging issues
+RUN pnpm install
```
Comment (roninjin10) [inline ops/docker/hardhat/Dockerfile:17]: Note we don't use frozen lockfile because there is no lockfile

File: package.json [hunk 18]
```diff
@@ -3,54 +3,25 @@
...
-      "**/@openzeppelin/*",
-      "@eth-optimism/contracts-periphery/ds-test",
-      "@eth-optimism/contracts-periphery/forge-std",
-      "@eth-optimism/contracts-periphery/@rari-capital/solmate",
-      "forta-agent"
-    ]
-  },
   "private": true,
+  "engines": {
+    "pnpm": ">=7"
```
Comment (roninjin10) [inline package.json:8]: would be a good idea to also add

"node": ">16",

but that is out of scope for this pr

Comment (roninjin10) [reply package.json:8]: this will make it so if user uses npm or yarn, everything will work fine but they will get a warning in console that pnpm is the correct engine to use

File: package.json [hunk 19]
```diff
@@ -3,54 +3,25 @@
...
+    "clean": "lerna run clean --parallel",
+    "build": "lerna run build",
+    "test": "lerna run test --parallel",
+    "test:coverage": "lerna run test:coverage --parallel",
+    "lint:ts:check": "lerna run lint:ts:check",
+    "lint:check": "lerna run lint:check",
+    "lint:fix": "lerna run lint:fix --parallel",
+    "lint:specs:fix": "markdownlint-cli2-fix \"./specs/**/*.md\"",
+    "lint:specs:check": "markdownlint-cli2  \"./specs/**/*.md\"",
+    "lint:specs:toc": "doctoc '--title=**Table of Contents**' ./specs",
```
Comment (roninjin10) [inline package.json:20]: Could nuke all this for pnpm

`pnpm concurrently run clean --parallel`

I wouldn't bother though because we really should use nx.  There is a pr that is pretty much ready to merge now https://github.com/ethereum-optimism/optimism/pull/4778

File: package.json [hunk 20]
```diff
@@ -76,7 +47,7 @@
     "eslint-plugin-react": "^7.24.0",
     "eslint-plugin-unicorn": "^42.0.0",
     "husky": "^6.0.0",
-    "lerna": "^4.0.0",
+    "lerna": "^6.0.0",
```
Comment (roninjin10) [inline package.json:50]: Upgrade lerna to version that supports pnpm

File: package.json [hunk 21]
```diff
@@ -3,54 +3,25 @@
...
+    "lint:fix": "lerna run lint:fix --parallel",
+    "lint:specs:fix": "markdownlint-cli2-fix \"./specs/**/*.md\"",
+    "lint:specs:check": "markdownlint-cli2  \"./specs/**/*.md\"",
+    "lint:specs:toc": "doctoc '--title=**Table of Contents**' ./specs",
     "postinstall": "patch-package",
-    "ready": "yarn lint && yarn test",
+    "ready": "pnpm lint && pnpm test",
     "prepare": "husky install",
-    "release": "yarn build && yarn changeset publish",
-    "update:yarn": "yarn set version 1.x"
```
Comment (roninjin10) [inline package.json:53]: This script was used to update yarn version in .yarn/releases

File: packages/atst/package.json [hunk 22]
```diff
@@ -26,13 +26,13 @@
...
+    "build": "tsup",
     "generate": "wagmi generate",
-    "lint": "yarn lint:fix && yarn lint:check",
+    "lint": "pnpm lint:fix && pnpm lint:check",
     "lint:check": "eslint . --max-warnings=0",
-    "lint:fix": "yarn lint:check --fix",
+    "lint:fix": "pnpm lint:check --fix",
     "test": "vitest",
-    "typecheck": "yarn tsc --noEmit"
+    "typecheck": "tsc --noEmit"
```
Comment (roninjin10) [inline packages/atst/package.json:35]: in pnpm (and yarn and npm) you don't need to type yarn or pnpm first to run a node_module script

File: .circleci/config.yml [hunk 23]
```diff
@@ -5,7 +5,7 @@ orbs:
   gcp-cli: circleci/gcp-cli@3.0.1
 commands:
   gcp-oidc-authenticate:
-    description: "Authenticate with GCP using a CircleCI OIDC token."
+    description: 'Authenticate with GCP using a CircleCI OIDC token.'
```
Comment (roninjin10) [inline .circleci/config.yml:8]: TODO my prettier/linter changed all these automatically

File: ops/docker/Dockerfile.packages (hunk 8 above)
Comment (roninjin10) [reply ops/docker/Dockerfile.packages:38]: Could likely get it to work on alpine but it would take a long time to figure out or a hack (simple hack is not running postinstall scripts at all which would work but is a hack and could cause hard to debug issues in future)

Comment (roninjin10) [reply ops/docker/Dockerfile.packages:38]: All node.js images should use a debian image by defaut.   If we needed an image as tiny and minimal as alpine, there are better alternatives such as copying into [gcr.io/distroless/nodejs18](https://github.com/roninjin10/stax/blob/main/Dockerfile#L242)

File: .npmrc (hunk 4 above)
Comment (maurelian) [reply .npmrc:10]: I didn't even know this file existed.

Comment (roninjin10) [reply .npmrc:10]: It didn't exist until this pr.   What I meant was the behavior pnpm has of never hoisting by default is what fixes a lot of pain you and Kelvin went through.  Anytime a type stopped working randomly or some dependency changed out from under you part of what caused that con fusing behavior was how yarn/npm hoists.

File: .npmrc [hunk 24]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (roninjin10) [inline .npmrc:1]:
```suggestion
```

File: .npmrc (hunk 3 above)
Comment (roninjin10) [reply .npmrc:3]: My nx pr removes all the prebuild pretest prelint ect. in favor of nx

File: packages/contracts-bedrock/package.json [hunk 25]
```diff
@@ -67,7 +67,7 @@
     "@foundry-rs/easy-foundryup": "^0.1.3",
     "@nomiclabs/hardhat-ethers": "^2.0.0",
     "@nomiclabs/hardhat-waffle": "^2.0.0",
-    "@rari-capital/solmate": "https://github.com/rari-capital/solmate.git#8f9b23f8838670afda0fd8983f2c41e8037ae6bc",
+    "@rari-capital/solmate": "https://github.com/transmissions11/solmate.git#8f9b23f8838670afda0fd8983f2c41e8037ae6bc",
```
Comment (roninjin10) [inline packages/contracts-bedrock/package.json:70]: Here I changed rari-capital to the repo that rari-capital redirects to if you go to https://github.com/rari-capital/solmate.git#8f9b23f8838670afda0fd8983f2c41e8037ae6bc .  This was causing an issue with pnpm correctly cloning submodules within solmate

File: .npmrc (hunk 3 above)
Comment (roninjin10) [reply .npmrc:3]: Not shamefully hoisting here means we need to test anything that doesn't have test coverage

**Comment by s7v7nislands** [comment]
@clabby any plan for this PR? Can I make a new PR to fix `lerna` too old, the node-gyp module can't compile on python 3.11?

**Comment by trianglesphere** [comment]
> Can I make a new PR to fix `lerna` too old, the node-gyp module can't compile on python 3.11?

Go for it. I think we likely want to do this migration, but it is not a top priority right now. This will also take a little bit of time b/c we don't want to break devs if/when we do this migration.
